firedrake/preconditioners/pmg.py

```diff
@@ -179,20 +194,40 @@ def inject_state(interpolators):
 
         cbcs = []
         for bc in fctx._problem.bcs:
-            # Don't actually need the value, since it's only used for
-            # killing parts of the matrix. This should be generalised
-            # for p-FAS, if anyone ever wants to do that
-
             cV_ = cV
             for index in bc._indices:
                 cV_ = cV_.sub(index)
 
-            cbcs.append(firedrake.DirichletBC(cV_, firedrake.zero(cV_.shape),
+            cbc_value = self.coarsen_bc_value(bc, cV_)
+            cbcs.append(firedrake.DirichletBC(cV_, cbc_value,
                                               bc.sub_domain,
                                               method=bc.method))
 
+        def get_degree(ele):
+            if isinstance(ele, (MixedElement, VectorElement, TensorElement, TensorProductElement)):
+                sub = ele.sub_elements()
+                return get_degree(sub[0])
+            else:
+                N = ele.degree()
+                try:
+                    N, = set(N)
+                except TypeError:
+                    pass
+                return N
+
+        # Coarsen the quadrature degree
+        # such that the ratio of GL to GLL nodes is preserved
         fcp = fctx._problem.form_compiler_parameters
-        cproblem = firedrake.NonlinearVariationalProblem(cF, cu, cbcs, cJ,
+        if fcp is not None:
```

Comment:
I think it's more usual to specify the quadrature degree by passing options to the `Measure` object. I don't think this will catch that case. Is that correct?

Reply:
In that case, whenever we specify the degree on the measure, could I just add a measure with the coarse degree in the replace dictionary?

Reply:
Does `ufl.replace' replace `Measure`s? I have never checked, but I wouldn't think so.

Reply:
It doesn't, no, measures are outside of that stuff.